### ISSUE TITLE
Generic `Type[T]` fails in unions for unions

### ISSUE THREAD

**Issue by brentyi** [microsoft/pyright#4298]
**Describe the bug**

I'm trying to write a function with a signature that looks like this:
```python
def func4(f: Type[T] | List[T]) -> T:
    ...
```
Which works fine with pyright when I pass in a class or list of integers, but fails when I try to pass in a union type (eg `func4(int | str)`):

```
reportGeneralTypeIssues:Argument of type "Type[int] | Type[str]" cannot be assigned to parameter "f" of type "Type[T@func4] | List[T@func4]" in function "func4" ××Type "object" cannot be assigned to type "Type[T@func4] | List[T@func4]"
```

This is potentially stretching the limits of what `typing.Type[]` is intended to be used for (particularly in the context of the `TypeForm` proposal https://github.com/python/mypy/issues/9773), but is suprising to me because `int | str` is compatible with `Type[T]` when applied outside of a union:
```python
def func2(f: Type[T]) -> T:
    ...


out2 = func2(int | str)
reveal_type(out2)  # int | str
```

**To Reproduce**

Here's an extended version of the snippets above, which highlights the error + some error-free relevant examples:

```python
from typing import Any, Callable, List, Tuple, Type, TypeVar

T = TypeVar("T")


# OK:


def func1(f: Type) -> Any:
    ...


out1 = func1(int | str)
reveal_type(out1)  # Any


# OK:


def func2(f: Type[T]) -> T:
    ...


out2 = func2(int | str)
reveal_type(out2)  # int | str


# OK:


def func3(f: Type | List) -> Any:
    ...


out3 = func3(int | str)
reveal_type(out3)  # Any


# Not ok!


def func4(f: Type[T] | List[T]) -> T:
    ...


out4a = func4([1, 2, 3])
reveal_type(out4a)  # int

out4b = func4(int | str)  # Error! Argument of type ... cannot be assigned to ...
reveal_type(out4b)  # Unknown
```

Thanks!!

**Comment by erictraut** [comment]
I agree that pyright's behavior is inconsistent here. I can see three options:
1. Leave the current behavior in place (allow `int | str` to match `type[T]` but not `type[T] | list[T]`).
2. Allow `int | str` to match `type[T]` and `type[T] | list[T]`.
3. Don't allow `int | str` to match `type[T]` or `type[T] | list[T]`.

Mypy opts for solution 3, and that's consistent with a strict reading of PEP 484. It also provides the most type safety. The reason pyright currently allows `int | str` to match `type[T]` is for practicality, even though it bends the rules a bit.

Option 2 widens the type safety gap, so that's probably not a good choice.

I'll need to think about this a bit more before deciding whether to stick with the current behavior (option 1) or switch to mypy's behavior (option 3).

**Comment by brentyi** [comment]
Thanks for the quick reply!

Some notes:

- (a) I agree with option 3 in principle, following the PEP seems like the correct thing to do, even if it's painful.
- (b) I would find a strict interpretation of the PEP very painful. I personally find the current behavior _super_ useful, and would celebrate internally if it was kept until `typing.TypeForm` happens (if it does happen).
- (c) If the decision is made to continue matching `int | str` with `type[T]`, it feels wrong to not match it with `type[T] | SomeOtherGeneric[T]`; why should taking the union between one type and another change the meaning of the first type? I can work around this by annotating with `type[T]` and `SomeOtherGeneric[T]` in separate overloads, but this looks a little bit weird.

To concretize (b), [the project](https://github.com/brentyi/tyro) I'm working on generates CLI interfaces from type annotations. Our interfaces are written with a `cli(target: Type[T]) -> T` function that is loosely used like:
```
# Construct a dataclass from command-line inputs.
instance_of_dataclass = cli(DataclassType)
```
or, sometimes:
```
# Construct one of two dataclass types.
dataclass_args = cli(DataclassTypeOne | DataclassTypeTwo)

# Construct a length-2 or length-3 tuple.
tuple_args = cli(tuple[int, int] | tuple[int, int, int])
```
where `cli()` generates an argument parser from the input type, then returns an instance of it.

This is being used in some projects with medium-sized userbases (eg ["nerfstudio"](https://github.com/nerfstudio-project/nerfstudio)), and it'd be nice to not break things for everyone. :slightly_smiling_face:

**Comment by Azureblade3808** [comment]
I'd like to see pyright go option 3 (mypy's behavior).

Except for `typing.cast`, I cannot recall any chance that a function will take a union type as its argument, which neither is instantiable nor supports `isinstance` or `issubclass` checks.

**Comment by Azureblade3808** [comment]
@brentyi Would you explain more about how you were going to implement your functions?

Let's take `func2` for an example. I imagine that it would be implemented with a call of the type's constructor, like --
```python
def func2(f: type[T]) -> T:
    return f()
```
But this doesn't work with `int | str`, right? 

What I can think of is using a typed variable as its argument --
```python
f: type[int | str] = ...
out2 = func2(f)
reveal_type(out2)  # "int | str"
```
But this also works with `func4` without being reported by pyright --
```
f: type[int | str] = ...
def func4(f: type[T] | list[T]) -> T: ...
out4 = func4(f)
reveal_type(out4)  # "int | str"
```

**Comment by brentyi** [comment]
> Except for `typing.cast`, I cannot recall any chance that a function will take a union type as its argument, which neither is instantiable nor supports `isinstance` or `issubclass` checks.

Pushback here:
- Not instantiable: neither are Protocols, but these match with `type`.
- `isinstance` and `issubclass`: as of PEP-0604, union types now work with both `isinstance` and `issubclass`; `isinstance(5, int | str)` will evaluate to `True`. Protocols also don't without `@runtime_checkable`, but still match with `type`.

Another standard library example where type forms like unions can be used as an argument is `dataclasses.Field()` and `dataclass.make_dataclass` ([typeshed link](https://github.com/python/typeshed/blob/main/stdlib/dataclasses.pyi)), which are both currently annotated using `type`.

> @brentyi Would you explain more about how you were going to implement your functions?

Sure; the function in question is actually already implemented in [this command-line parsing library](https://github.com/brentyi/tyro).

The implementation is quite long but here's a working example:
```python
# script.py

from tyro import cli  # (target: Type[T]) -> T

user_input = cli(int | str | tuple[int, str, int])
print(f"{user_input=} {type(user_input)=}")
```

And some real outputs when I run `script.py`:

```python
$ python script.py 0
user_input=0 type(user_input)=<class 'int'>

$ python script.py zero
user_input='zero' type(user_input)=<class 'str'>

$ python script.py 1 2 3
user_input=(1, '2', 3) type(user_input)=<class 'tuple'>
```

**Comment by hmc-cs-mdrissi** [comment]
I similarly have a library that really likes unions matching type[T]. It’s a library that analyzes runtime type annotations to derive serialization and deserialization logic similar to Cattrs. For union type when deserializing the serialization format is designed to always have enough information to recognize which type in union it came from. For some times it’s able to tell just by structure of it while in more ambiguous cases the serialized file includes info of which type in union it came from. The api is designed to handle,

config = to_config(ty, obj)
obj2 = from_config(ty, config)

For any type registered in library which is roughly basic types (including generic containers like dict/list/tuple/set) + a mechanism to define serializable types that either requires the class’s constructor/fields to all have type hints of already known serializable types or you register a parser/serializer for any complex case.

This example is very similar to the library idea behind type form pep. Runtime type introspection can handle unions and other forms.

**Comment by Azureblade3808** [comment]
> `isinstance` and `issubclass`: as of PEP-0604, union types now work with both `isinstance` and `issubclass`; `isinstance(5, int | str)` will evaluate to `True`. Protocols also don't without `@runtime_checkable`, but still match with `type`.

Good point. But a union type still fails `isinstance(union_type, type)` (not so important) and doesn't hold some of `type`'s members (important). One may naturally think that `type[T]` would be a `type` subclass and get supprised if some member access fails on a union type which is passed as an sargument. On the contrary, protocols are just real `type` subclasses.

I do think python needs something that distinguishes a conceptual type from a real class though.

**Comment by erictraut** [comment]
After considering all of the feedback, I've decided to leave the current implementation as-is. We can revisit this again if and when `TypeForm` is introduced to the type system.
